backend/wmg/pipeline/summary_cubes/expression_summary_fmg/transform.py

```diff
@@ -0,0 +1,123 @@
+import concurrent
+import logging
+
+import numba as nb
+import numpy as np
+import pandas as pd
+
+import tiledb
+
+from backend.common.utils.math_utils import MB
+from backend.wmg.pipeline.summary_cubes.extract import extract_obs_data
+
+from backend.wmg.data.schemas.corpus_schema import INTEGRATED_ARRAY_NAME
+from backend.wmg.data.tiledb import create_ctx
+from backend.wmg.data.utils import log_func_runtime
+
+logger = logging.getLogger(__name__)
+
+BINOMIAL_NNZ_RANKIT_THR = 1
+
+
+def transform(
+    corpus_path: str, gene_ontology_term_ids: list, cube_dims: list
+) -> (pd.DataFrame, np.ndarray, np.ndarray):
+    """
+    Build the summary cube with rankit expression sum & sum of squares, nnz
+    (num cells with non zero expression) values for each gene for each possible
+    group of cell attributes (cube row).
+    """
+
+    cell_labels, cube_index = make_cube_index(corpus_path, cube_dims)
+    n_groups = len(cube_index)
+    n_genes = len(gene_ontology_term_ids)
+
+    cube_sum = np.zeros((n_groups, n_genes), dtype=np.float32)
+    cube_sqsum = np.zeros((n_groups, n_genes), dtype=np.float32)
+    cube_nnz = np.zeros((n_groups, n_genes), dtype=np.uint64)
+    cube_nnz_thr = np.zeros((n_groups, n_genes), dtype=np.uint64)
+
+    reduce_X(corpus_path, cell_labels.cube_idx.values, cube_sum, cube_sqsum, cube_nnz, cube_nnz_thr)
+    return cube_index, cube_sum, cube_sqsum, cube_nnz, cube_nnz_thr
+
+
+@log_func_runtime
+def reduce_X(
+    tdb_group: str,
+    cube_indices: np.ndarray,
+    cube_sum: np.ndarray,
+    cube_sqsum: np.ndarray,
+    cube_nnz: np.ndarray,
+    cube_nnz_thr: np.ndarray,
+):
+    """
+    Reduce the expression data stored in the integrated corpus by summing it by gene for each cube row (unique combo
+    of cell attributes)
+    """
+    with concurrent.futures.ThreadPoolExecutor() as executor:
+        cfg = {
+            "py.init_buffer_bytes": 512 * MB,
+            "py.exact_init_buffer_bytes": "true",
+        }
+        with tiledb.open(f"{tdb_group}/{INTEGRATED_ARRAY_NAME}", ctx=create_ctx(config_overrides=cfg)) as expression:
+            iterable = expression.query(return_incomplete=True, order="U", attrs=["rankit"])
+            future = None
+            for i, result in enumerate(iterable.df[:]):
+                logger.info(f"reduce integrated expression data, iter {i}")
+                if future is not None:
+                    future.result()  # forces a wait
+                future = executor.submit(
```

Comment:
Is this actually executing in parallel? It looks like we are waiting for the previous thread to return before starting the next. Are there some dependencies between the threads that prevent them from running at the same time?

Reply:
I'm not exactly sure if this is going row-by-row in the target array for the reduction. It looks like it might not be. In that case, we want the wait to avoid the race condition. I'm not sure why it's being executed in a thread-pool to begin with.